### PR TITLE
No_Issue? - Code Review for the Section Tool

### PR DESCRIPTION
**Pull request by pushyamig**
Reviewers: @bkirschn @zqian @dovek @dlhaines @lsloan

### CODE REVIEW

File: sectionsUtilityTool/src/main/java/edu/umich/ctools/sectionsUtilityTool/SectionsUtilityToolServlet.java [hunk 1]
```diff
@@ -15,93 +15,131 @@
...
 	}
+
 	
-	private String getCanvasAdminToken() {
-		return "REPLACE WITH ACTUAL TOKEN";
+	protected void getCanvasCredentials() {
+		M_log.debug("getCanvasCredentials(): called");
+		String propertiesFilePath = System.getProperty(SYSTEM_PROPERTY_FILE_PATH);
+		if (!isEmpty(propertiesFilePath)) {
+		canvasProperties=PropertiesUtilities.getPropertiesObjectFromURL(propertiesFilePath);
```
Comment (dlhaines) [inline sectionsUtilityTool/src/main/java/edu/umich/ctools/sectionsUtilityTool/SectionsUtilityToolServlet.java]: I'd make a distinction between canvasProperties and canvasCredentials.  We should assume that credentials will be kept in a separate file in a separate directory.

Comment (pushyamig) [inline sectionsUtilityTool/src/main/java/edu/umich/ctools/sectionsUtilityTool/SectionsUtilityToolServlet.java]: I need to use better name. i will clean up that. 

I have a external properties file some where on linux box in a directory, and the properties file has list of properties like 
canvas.token=asbjdbs
canvas.url =http://test.canvas.com
use.test.url =true 
ldap.server.name = ldap.server.umich.edu

why do need to place the just canvas related properties in another separate file?

Comment (pushyamig) [inline sectionsUtilityTool/src/main/java/edu/umich/ctools/sectionsUtilityTool/SectionsUtilityToolServlet.java]: I have added 2 properties files and 2 -D's to look up those file

File: sectionsUtilityTool/src/main/webapp/assets/js/filters/filters.js [hunk 2]
```diff
@@ -1,12 +1,34 @@
...
+	  	    }
+      }
+    }
+    return filtered;
   };
-}).filter('anotherFilter', function () {
-  return function () {
+}).filter('showOnlyTermsWithCourses', function () {
+  return function (terms) {
+    var termsArray = [39,80,1];
```
Comment (zqian) [inline sectionsUtilityTool/src/main/webapp/assets/js/filters/filters.js]: any special meanings of number 39, 80, or 1 in the array?

Comment (gsilver) [inline sectionsUtilityTool/src/main/webapp/assets/js/filters/filters.js]: Hai - this filter is not used. Neither is 'anotherFilter'

Will need some severe clean up soon.

File: sectionsUtilityTool/README.txt [hunk 3]
```diff
@@ -0,0 +1,18 @@
+[Sections Utility Tool]
```
Comment (zqian) [inline sectionsUtilityTool/README.txt]: You may want to rename this file to README.md, to take advantage of Markdown syntax.

**Comment by lsloan** [comment]
Remember:  To call our attention to pull requests for code reviews, include in the request's description (you didn't give any) the GitHub usernames of all the people you want to participate.  All of those people should receive an email notification from GitHub.

Just write: @lsloan @dlhaines @gsilver @zqian @dovek @bkirschn

File: SIS_import/sis_upload.rb [hunk 4]
```diff
@@ -4,93 +4,151 @@
...
+	p server
 
-		if (parsed["errors"])
-			## break and print error
-			error_array=parsed["errors"]
-			## hashmap ["message"=>"error_message"
-			uploadError = error_array[0]["message"]
-			outputFile.write("upload error: " + uploadError)
-			outputFile.write("\n")
+	json_data=`curl -H "Content-Type: application/zip" --data-binary @#{fileName} -H "Authorization: Bearer #{token}" #{server}/api/v1/accounts/1/sis_imports.json?import_type=instructure_csv`
```
Comment (lsloan) [inline SIS_import/sis_upload.rb]: I believe I may have mentioned this before.  Calling curl in this way is not a very good idea.  It adds an external dependency to the application.  This call also starts up another shell which will then run curl, which makes for at least two additional processes, maybe more.  Those processes could die or be terminated, which means nothing will be returned to this program.  However, the value of json_data isn't checked at all.  The call to curl in another shell should be replaced with a call to a native Ruby HTTP client.  If that's done, there will probably no longer be a need to write data to a file to pass to the HTTP client.  It will probably accept the data as part of the function/method call.

File: SIS_import/sis_upload.rb [hunk 5]
```diff
@@ -4,93 +4,151 @@
...
+		#sleep every 10 sec, before checking the status again
+		sleep(10);
 
-		if (parsed_result["errors"])
-				## break and print error
-				uploadError=parsed_result["errors"]
-				## hashmap ["message"=>"error_message"
-				outputFile.write("upload error: " + uploadError)
-				outputFile.write("\n")
+		json_result=`curl '#{server}/api/v1/accounts/1/sis_imports/#{job_id}' -H "Authorization: Bearer #{token}"`
```
Comment (lsloan) [inline SIS_import/sis_upload.rb]: Another unchecked call to curl that should be replaced.

File: SIS_import/sis_upload.rb [hunk 6]
```diff
@@ -4,93 +4,151 @@
...
+		p "found recent id file #{id_file}"
+		process_id = ''
+		File.open(id_file, 'r') do |idFile|
+			while line = idFile.gets
+				# only read the first line, which is the token value
+				process_id=line.strip
+				break
+			end
+		end
+		process_result=`curl '#{server}/api/v1/accounts/1/sis_imports/#{process_id}' -H "Authorization: Bearer #{token}"`
```
Comment (lsloan) [inline SIS_import/sis_upload.rb]: Another unchecked call to curl that should be replaced.

File: SIS_import/sis_upload.rb [hunk 7]
```diff
@@ -154,69 +249,92 @@ def parseJson(data)
...
-	p "SIS upload finished with " + fileName
-	exit
-else
 	## check first about the environment variable setting for MAILTO '
-	p " use the environment variable for sending out error messages #{ENV['MAILTO']}"
+	p "Use the environment variable 'MAILTO' for sending out error messages to #{ENV['MAILTO']}"
 	## send email to support team with the error message
-	`echo #{uploadError} | mail -s "#{server} Upload Error" #{ENV['MAILTO']}`
-	abort(uploadError)
+	`echo #{upload_error} | mail -s "#{server} Upload Error" #{ENV['MAILTO']}`
```
Comment (lsloan) [inline SIS_import/sis_upload.rb]: Just as Ruby offers HTTP clients, it also offers SMTP clients.  Replace this external call to mail with a function call to a SMTP function/method.

File: SIS_import/sis_upload.rb [hunk 8]
```diff
@@ -154,69 +249,92 @@ def parseJson(data)
...
+		end
+
+		if (upload_error)
+			# write upload error string into file
+			outputFile.write(upload_error)
+
+			## check first about the environment variable setting for MAILTO '
+			p "Use the environment variable 'MAILTO' for sending out error messages to #{ENV['MAILTO']}"
+			## send email to support team with the error message
+			`echo #{upload_error} | mail -s "#{server} Upload Error" #{ENV['MAILTO']}`
```
Comment (lsloan) [inline SIS_import/sis_upload.rb]: Just as Ruby offers HTTP clients, it also offers SMTP clients.  Replace this external call to mail with a function call to a SMTP function/method.

File: SIS_import/sis_upload.rb (hunk 4 above)
Comment (zqian) [inline SIS_import/sis_upload.rb]: This is in my SIS upload project, not related to Pushyami's Section Management Tool. 

I have replaced the curl call in sis_update_url.rb file, but have not changed the sis_upload.rb yet. I will do that shortly, and open a separate Pull Request for it.

File: SIS_import/sis_upload.rb (hunk 5 above)
Comment (zqian) [inline SIS_import/sis_upload.rb]: See above comments.

File: SIS_import/sis_upload.rb (hunk 6 above)
Comment (zqian) [inline SIS_import/sis_upload.rb]: See above comments.

File: SIS_import/sis_upload.rb (hunk 7 above)
Comment (zqian) [inline SIS_import/sis_upload.rb]: Will address.

Comment (lsloan) [inline SIS_import/sis_upload.rb]: Net::SMTP looks helpful.

File: SIS_import/sis_upload.rb (hunk 4 above)
Comment (lsloan) [inline SIS_import/sis_upload.rb]: Net::HTTP would be helpful to you.

File: sectionsUtilityTool/src/main/java/edu/umich/ctools/sectionsUtilityTool/SectionsUtilityToolServlet.java [hunk 9]
```diff
@@ -15,93 +15,131 @@
...
-		System.out.println("HeaderNames: "+headerNames);
-		System.out.println("RequestURI: "+requestURI);
-		System.out.println("QueryString: "+queryString);
-		System.out.println("ContextPath: "+contextPath);
 	}
+
 	
-	private String getCanvasAdminToken() {
-		return "REPLACE WITH ACTUAL TOKEN";
+	protected void getCanvasCredentials() {
```
Comment (dlhaines) [inline sectionsUtilityTool/src/main/java/edu/umich/ctools/sectionsUtilityTool/SectionsUtilityToolServlet.java]: If this returned the properties object rather than set a global variable it would be easier to test and debug.

File: sectionsUtilityTool/src/main/webapp/section_data/api-list.txt [hunk 10]
```diff
@@ -0,0 +1,11 @@
...
+
+https://umich.instructure.com/api/v1/courses?as_user_id=sis_login_id:<uniqname>&per_page=100&enrollment_term_id=sis_term_id:<term-id>&published=true&with_enrollments=true&enrollment_type=teacher&access_token=<acccess-token>
+
+Get sections
+
+https://umich.instructure.com/api/v1/courses/<course-id>/sections?per_page=100&access_token=<acccess-token>
+
+Get terms
+
+https://umich.instructure.com/api/v1/accounts/1/terms
```
Comment (lsloan) [inline sectionsUtilityTool/src/main/webapp/section_data/api-list.txt]: What's the purpose of this file?

File: sectionsUtilityTool/src/main/webapp/section_data/courses-amanjo.json [hunk 11]
```diff
@@ -0,0 +1 @@
...ic_syllabus":false,"storage_quota_mb":1000,"apply_assignment_group_weights":false,"calendar":{"ics":"https://umich.instructure.com/feeds/calendars/course_b4ENPXPOOnSHLNeDa1apKXmRs0iTAZh2mzg8hr0s.ics"},"sis_course_id":"20200116962","integration_id":null,"enrollments":[{"type":"teacher","role":"TeacherEnrollment","role_id":15,"enrollment_state":"active"},{"type":"teacher","role":"TeacherEnrollment","role_id":15,"enrollment_state":"active"}],"hide_final_grades":false,"workflow_state":"available"},{"account_id":281,"course_code":"THTREMUS 340","default_view":"syllabus","enrollment_term_id":39,"id":610,"is_public":null,"name":"Devising Theatre","start_at":"2015-01-07T20:18:23Z","end_at":null,"public_syllabus":false,"storage_quota_mb":1000,"apply_assignment_group_weights":false,"calendar":{"ics":"https://umich.instructure.com/feeds/calendars/course_6p4cpJLVjRGYIbkqSaKWavcP5XefumBhZcjCx4iz.ics"},"sis_course_id":"20200388391","integration_id":null,"enrollments":[{"type":"teacher","role":"TeacherEnrollment","role_id":15,"enrollment_state":"active"},{"type":"teacher","role":"TeacherEnrollment","role_id":15,"enrollment_state":"active"}],"hide_final_grades":false,"workflow_state":"available"}]
```
Comment (lsloan) [inline sectionsUtilityTool/src/main/webapp/section_data/courses-amanjo.json]: I'm surprised by the large number of JSON files of course section data.  If the courses change, it means that this data needs to be changed and redeployed, right?  Why isn't this data stored in a DB?

File: sectionsUtilityTool/src/main/webapp/assets/vendor/angular/sortable.js [hunk 12]
```diff
@@ -0,0 +1,350 @@
...
+          } else {
+            $log.info('ui.sortable: ngModel not provided!', element);
+          }
+
+          // Create sortable
+          element.sortable(opts);
+        }
+      };
+    }
+  ]);
```
Comment (lsloan) [inline sectionsUtilityTool/src/main/webapp/assets/vendor/angular/sortable.js]: The vendor assets really clutter up the code review.  To make it easier to see what needs to be reviewed and what doesn't, things like these should be put into a separate pull request.  Probably an earlier one.

File: sectionsUtilityTool/src/main/webapp/assets/js/utils.js [hunk 13]
```diff
@@ -30,6 +28,54 @@ var errorHandler = function (url, result) {
   return errorResponse;       
 };
 
+var getCurrentTerm = function(termData) {
+  var now = moment();
+  var currentTerm = [];
+  $.each(termData, function() {
+    //horrifying
```
Comment (lsloan) [inline sectionsUtilityTool/src/main/webapp/assets/js/utils.js]: :laughing:

File: sectionsUtilityTool/src/main/webapp/assets/js/utils.js [hunk 14]
```diff
@@ -30,6 +28,54 @@ var errorHandler = function (url, result) {
...
+    //TODO: deal with Med School terms later - as this is filtering anyth8ing that does not end in '0'
+    if(moment(this.start_at).isBefore(now) && moment(this.end_at).isAfter(now)) {
+      if (this.sis_term_id !== null && this.sis_term_id !== undefined  && this.sis_term_id.slice(-1) ==='0'){
+        currentTerm.currentTermId =  this.sis_term_id;
+        currentTerm.currentTermName =  this.name;
+        currentTerm.currentTermCanvasId =  this.id;
+      }
```
Comment (lsloan) [inline sectionsUtilityTool/src/main/webapp/assets/js/utils.js]: Philosophical discussion: Maybe break out of the loop here once the current term has been found so that the entire remaining list doesn't need to be processed.  For example, what if the list were very long and the one we needed was the first one?  There may be a delay while the others are evaluated.

**Comment by lsloan** [comment]
Done with my review of this PR.  Move my post-it note, please.  :smiley:

File: sectionsUtilityTool/src/main/java/edu/umich/ctools/sectionsUtilityTool/SectionsUtilityToolServlet.java [hunk 15]
```diff
@@ -15,93 +15,131 @@
...
+	}
+	protected void doPut(HttpServletRequest request,HttpServletResponse response) throws IOException {
+		M_log.debug("doPut: Called");
+		canvasRestApiCall(request, response);
+		
+	}
+	
+	
+
+	private void canvasRestApiCall(HttpServletRequest request,
```
Comment (bkirschn) [inline sectionsUtilityTool/src/main/java/edu/umich/ctools/sectionsUtilityTool/SectionsUtilityToolServlet.java]: A method comment would be helpful here -- this method actually handles _any_ get, post, or put request (which isn't clear from the method name)

File: sectionsUtilityTool/src/main/java/edu/umich/ctools/sectionsUtilityTool/SectionsUtilityToolServlet.java [hunk 16]
```diff
@@ -15,93 +15,131 @@
...
+		}
+		HttpClient client = new DefaultHttpClient();
 		final ArrayList<NameValuePair> nameValues = new ArrayList<NameValuePair>();
-	    nameValues.add(new BasicNameValuePair("Authorization", "Bearer"+ " " +getCanvasAdminToken()));
-	    nameValues.add(new BasicNameValuePair("content-type", "application/json"));
-	    for (final NameValuePair h : nameValues)
-	    {
-	        clientRequest.addHeader(h.getName(), h.getValue());
-	    }
+		nameValues.add(new BasicNameValuePair("Authorization", "Bearer"+ " " +canvasToken));
```
Comment (bkirschn) [inline sectionsUtilityTool/src/main/java/edu/umich/ctools/sectionsUtilityTool/SectionsUtilityToolServlet.java]: A comment here would be good, since we're using a special admin account to handle all canvas requests for any user

File: sectionsUtilityTool/src/main/webapp/section_data/courses-amanjo.json (hunk 11 above)
Comment (lsloan) [inline sectionsUtilityTool/src/main/webapp/section_data/courses-amanjo.json]: Yesterday David told me that these JSON files are actually for testing, not production use.  They should be moved into a directory called "tests" to make that clear.

File: sectionsUtilityTool/README.txt (hunk 3 above)
Comment (pushyamig) [inline sectionsUtilityTool/README.txt]: Done

File: sectionsUtilityTool/src/main/java/edu/umich/ctools/sectionsUtilityTool/SectionsUtilityToolServlet.java [hunk 17]
```diff
@@ -15,94 +15,120 @@
...
-		else {
-			url=targetURI+pathInfo;
-			//url=servletPath;
+		url= canvasURL+pathInfo+"?"+queryString;
+		 }else {
+			url=canvasURL+pathInfo;
+		}
+		String sessionId = request.getSession().getId();
+		StringBuilder loggingApiWithSessionInfo =new StringBuilder();
+		loggingApiWithSessionInfo.append("Canvas API request with Session Id:  \"");
```
Comment (dlhaines) [inline sectionsUtilityTool/src/main/java/edu/umich/ctools/sectionsUtilityTool/SectionsUtilityToolServlet.java]: String.format could be used to get rid of a lot of appends and tighten up the code.

**Comment by bkirschn** [comment]
No longer needed - pull request created only for code review
